### PR TITLE
WIP: Socket.io v3

### PR DESCRIPTION
**Pull request by HMarzban**

### CODE REVIEW

**Comment by JohnMcLear** [comment]
Excited to watch the tests run 👯‍♀️

File: src/node/handler/PadMessageHandler.js [hunk 1]
```diff
@@ -112,13 +112,13 @@ exports.handleConnect = (socket) => {
...
 exports.kickSessionsFromPad = (padID) => {
-  if (typeof socketio.sockets.clients !== 'function') return;
+  if (typeof socketio.sockets !== 'object') return;
 
   // skip if there is nobody on this pad
   if (_getRoomSockets(padID).length === 0) return;
 
   // disconnect everyone from this pad
-  socketio.sockets.in(padID).json.send({disconnect: 'deleted'});
+  socketio.in(padID).emit("message", {disconnect: 'deleted'});
```
Comment (JohnMcLear) [inline src/node/handler/PadMessageHandler.js:121]: lint will fail on this, expecting to see ``'message'`` not ``"message"``

File: src/package.json [hunk 2]
```diff
@@ -243,7 +244,7 @@
   },
   "scripts": {
     "lint": "eslint .",
-    "test": "mocha --timeout 120000 --recursive tests/backend/specs ../node_modules/ep_*/static/tests/backend/specs",
+    "test": "mocha --timeout 12000000 --recursive tests/backend/specs ../node_modules/ep_*/static/tests/backend/specs",
```
Comment (JohnMcLear) [inline src/package.json:247]: Why such a large timeout increase?

File: src/static/js/collab_client.js [hunk 3]
```diff
@@ -178,10 +178,11 @@ const getCollabClient = (ace2editor, serverVars, initialUserInfo, options, _pad)
   };
 
   const sendMessage = (msg) => {
-    getSocket().json.send(
+    getSocket().emit("message",
```
Comment (JohnMcLear) [inline src/static/js/collab_client.js:181]: ditto eslint issue :) -" +'

File: src/templates/pad.html [hunk 4]
```diff
@@ -170,7 +170,7 @@ <h2 data-l10n-id="pad.settings.myView"></h2>
 
           <h2 data-l10n-id="pad.settings.about">About</h2>
           <span data-l10n-id="pad.settings.poweredBy">Powered by</span>
-          <a href="https://etherpad.org">Etherpad</a>
+          <a href="https://etherpad.org">Etherpad-lite</a>
```
Comment (JohnMcLear) [inline src/templates/pad.html:173]: This should not be here :)

File: src/templates/pad.html [hunk 5]
```diff
@@ -464,7 +464,7 @@ <h1>Skin Builder</h1>
       // @license-end
   </script>
 
-  <script type="text/javascript" src="../socket.io/socket.io.js?v=<%=settings.randomVersionString%>"></script>
+  <script type="text/javascript" src="../socket.io/socket.io.js"></script>
```
Comment (JohnMcLear) [inline src/templates/pad.html:467]: This should not be here :)

File: src/tests/backend/common.js [hunk 6]
```diff
@@ -32,14 +32,14 @@ exports.init = async function () {
...
   agentPromise = new Promise((resolve) => { agentResolve = resolve; });
-
-  if (!logLevel.isLessThanOrEqualTo(log4js.levels.DEBUG)) {
-    exports.logger.warn('Disabling non-test logging for the duration of the test. ' +
-                        'To enable non-test logging, change the loglevel setting to DEBUG.');
-    log4js.setGlobalLogLevel(log4js.levels.OFF);
-    exports.logger.setLevel(logLevel);
-  }
-
+
```
Comment (JohnMcLear) [inline src/tests/backend/common.js:35]: This should not be here :)

File: src/tests/backend/specs/socketio.js [hunk 7]
```diff
@@ -54,7 +54,7 @@ const connect = async (res) => {
   logger.debug('socket.io connecting...');
   const socket = io(`${common.baseUrl}/`, {
     forceNew: true, // Different tests will have different query parameters.
-    path: '/socket.io',
+    // path: '/socket.io',
```
Comment (JohnMcLear) [inline src/tests/backend/specs/socketio.js:57]: don't comment out, just remove.

We do use /socket.io path for query string but we can modify that no problem.

File: src/tests/backend/specs/socketio.js [hunk 8]
```diff
@@ -104,7 +104,7 @@ describe(__filename, function () {
   };
   let socket;
 
-  before(async function () { agent = await common.init(); });
+  before(async function () {agent = await common.init(); });
```
Comment (JohnMcLear) [inline src/tests/backend/specs/socketio.js:107]: This should not be here :)

File: src/tests/backend/specs/socketio.js [hunk 9]
```diff
@@ -194,7 +194,6 @@ describe(__filename, function () {
       this.timeout(400);
       settings.requireAuthentication = true;
       const res = await agent.get('/p/pad').expect(401);
-      // Despite the 401, try to create the pad via a socket.io connection anyway.
```
Comment (JohnMcLear) [inline src/tests/backend/specs/socketio.js:197]: Why is this removed?

**Review by JohnMcLear — COMMENTED** [review]
Btw, you can use "draft" pull requests for WIP things :)

**Comment by HMarzban** [comment]
If `setting.loadTest: true`, in most cases the test will fail! ( `webaccess` issue)
The `userCanModify` and `checkAccess` functions in `/hooks/express/webaccess.js` file,  need to modify!

Also, `websocket`, `chat`, etc. tests passed successfully! But on the client side, when the user makes new changes, the other client will not receive new changes! (This is the main issue right now)

**Comment by HMarzban** [comment]
From now on, we have to set the `transports` option on the client and server to `io.connect (opt)`!
(This branch; `src/node/hooks/express/socketio.js:73`)

In this case we must have to put `settings.socketTransportProtocols` in `clientVars` for "pad.html" (and where necessary) when it's render in the back-end.

File: src/templates/pad.html (hunk 5 above)
Comment (HMarzban) [reply src/templates/pad.html:467]: Unfortunately I had to remove the random version tag by force, otherwise the socket.io client script would not load!

File: src/package.json (hunk 2 above)
Comment (HMarzban) [reply src/package.json]: my bad

**Comment by JohnMcLear** [comment]
> 
> 
> If `setting.loadTest: true`, in most cases the test will fail! ( `webaccess` issue)
> The `userCanModify` and `checkAccess` functions in `/hooks/express/webaccess.js` file, need to modify!
> 
> Also, `websocket`, `chat`, etc. tests passed successfully! But on the client side, when the user makes new changes, the other client will not receive new changes! (This is the main issue right now)

This is because the tests mostly don't cover real time collaboration, this is something I'm working on at the moment :)

**Comment by HMarzban** [comment]
Well, let me drop the mic, the real time collaboration issue was resolved. work like magic ⚔🧝‍♂️

**Comment by JohnMcLear** [comment]
This works enough for me to run the load tests, and it does appear that the impact on Etherpad is negligible so this wont help much in achieving additional authors on a pad.  

Our next step should be:
1. Confirming it's still socketIO creating the top load (I'm 99% sure it will but I'd like fresh eyes to help profile).
2. Look to migrate to websocket and drop socketio.

My conclusion here is that SocketIO is more of a burden now than a gain.

File: src/node/handler/PadMessageHandler.js [hunk 10]
```diff
@@ -951,6 +949,9 @@ const handleClientReady = async (socket, message, authorID) => {
       padIds.readonly || !webaccess.userCanModify(message.padId, socket.client.request);
 
   const {session: {user} = {}} = socket.client.request;
+  const {accessStatus} = await securityManager
+      .checkAccess(sessioninfos[socket.id].auth.padID, sessioninfos[socket.id].auth.sessionID, sessioninfos[socket.id].auth.token, user);
+
```
Comment (webzwo0i) [inline src/node/handler/PadMessageHandler.js:954]: Can you explain why this is needed here?

File: src/node/handler/PadMessageHandler.js [hunk 11]
```diff
@@ -1061,6 +1062,7 @@ const handleClientReady = async (socket, message, authorID) => {
         rev: pad.getHeadRevisionNumber(),
         time: currentTime,
       },
+      accessStatus,
```
Comment (webzwo0i) [inline src/node/handler/PadMessageHandler.js:1065]: Can you explain why this is here? CLIENT_READY should iirc not be sent when permissions are not sufficient to edit/join the pad?

File: src/static/js/collab_client.js [hunk 12]
```diff
@@ -178,10 +178,11 @@ const getCollabClient = (ace2editor, serverVars, initialUserInfo, options, _pad)
   };
 
   const sendMessage = (msg) => {
-    getSocket().json.send(
+    getSocket().emit('message',
         {
           type: 'COLLABROOM',
           component: 'pad',
+          padId: clientVars.padId,
```
Comment (webzwo0i) [inline src/static/js/collab_client.js:185]: Can you explain why this is needed?

**Comment by JohnMcLear** [comment]
Note that we will need to ensure padId is still kept within the query string as per https://github.com/ether/etherpad-lite/pull/4793

File: src/node/handler/PadMessageHandler.js (hunk 11 above)
Comment (HMarzban) [reply src/node/handler/PadMessageHandler.js]: In the first round of tests I needed it, it is not necessary, I removed it.

File: src/static/js/collab_client.js (hunk 12 above)
Comment (HMarzban) [reply src/static/js/collab_client.js]: As above, in the first round `usersession` does not work properly and I have to send` padId` from the client

File: src/node/handler/PadMessageHandler.js (hunk 10 above)
Comment (HMarzban) [reply src/node/handler/PadMessageHandler.js:954]: It return to the previous form.

**Comment by HMarzban** [comment]
I was reading this changeset recently:
https://socket.io/blog/monthly-update-3/#Socket-IO-v4

If we do this migration, The road is paved for us for further upgrades

**Comment by HMarzban** [comment]
@webzwo0i Sorry to be late for comments, I was in vacation.

**Comment by HMarzban** [comment]
Hey guys, have you thought about this PR! I have already fixed all the problems related to this upgrade! do you have a socket upgrade plan or is it necessary anymore?
@JohnMcLear 
@webzwo0i

**Comment by rhansen** [comment]
Before we can merge this we need to figure out a way to support plugins that expect socket.io 2.x. For example: https://github.com/ether/ep_webrtc/blob/7e5941f4a21ab869a8448ecb771c844643f3ef94/index.js#L46-L48

**Comment by HMarzban** [comment]
It's okay, I can help you! I change this function like below and it works like a charm!

```js
const handleRTCMessage = (socket, client, payload) => {
  // if(!socketIo) return false
  const userId = payload.from;
  const padId = payload.padId;
  const to = payload.to;

  const msg = {
    type: 'COLLABROOM',
    data: {
      type: 'RTC_MESSAGE',
      payload: {
        from: userId,
        to,
        data: payload.data,
      },
    },
  };
  socketIo.to(padId).emit('RTC_MESSAGE', msg);
};
```

**Comment by rhansen** [comment]
My point is that we need to audit all plugins to determine which ones depend on v2-specific behavior, then update them to work both with and without this PR. Alternatively, change this PR so that the plugins that depend on v2-specific behavior somehow continue to work unchanged.

**Comment by SamTV12345** [comment]
@HMarzban I know it's been some time since you  worked on this pull request. During the last 2.5 years has changed and I am currently the only maintainer of Etherpad. I am looking for help with the upgrade to socket.io  v4. So if you find some time it would be awesome if you could help me with the upgrade. 
With the help of @JohnMcLear I managed to revive Etherpad, made lots of changes to Etherpad and introduced Typescript to ueberdb. If you prepare a pull request you can be sure that it will be merged. I normally answer in at most one day. 

So TDLR: It would be awesome if you could help me with the socket io v4 upgrade.

**Comment by HMarzban** [comment]
Hi, @SamTV12345

Thank you for reaching out and for keeping Etherpad thriving! It's great to hear about the progress and the introduction of TypeScript to ueberdb.

I'm excited to help with the socket.io v4 upgrade. 
_(It was a quick, challenging experience last time, but I did it and it worked!)_ 

However, I'm currently swamped with tasks and will be busy until the 2nd of February. But right after that, I'm all in! 
I'm looking forward to diving back into the **wavy**, **turbulent** ocean of Etherpad code and exploring the best paths for the upgrade.
_**(I appreciate your patience.)**_

**Comment by SamTV12345** [comment]
> Hi, @SamTV12345
> 
> Thank you for reaching out and for keeping Etherpad thriving! It's great to hear about the progress and the introduction of TypeScript to ueberdb.
> 
> I'm excited to help with the socket.io v4 upgrade. _(It was a quick, challenging experience last time, but I did it and it worked!)_
> 
> However, I'm currently swamped with tasks and will be busy until the 2nd of February. But right after that, I'm all in! I'm looking forward to diving back into the **wavy**, **turbulent** ocean of Etherpad code and exploring the best paths for the upgrade. _**(I appreciate your patience.)**_

Thanks for the help. This is very much appreciated. Yeah the Etherpad code with its hundreds of public methods is a bit overwhelming at least for me. So any help is awesome. Sure it is not a problem if you are currently busy. So until February and thanks for reporting back.

**Comment by SamTV12345** [comment]
I guess we can finally close this <3
